File: COA software/backend/app/schemas/ai_planner.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, field_validator


class OptimizationObjectiveWeights(BaseModel):
    asset_availability: float = Field(default=40.0, description="Weight for asset availability gain")
    maintenance_priority: float = Field(default=25.0, description="Weight for maintenance priority coverage")
    train_impact: float = Field(default=20.0, description="Weight for minimizing train timetable delay")
    block_utilization: float = Field(default=15.0, description="Weight for block time utilization efficiency")

    @field_validator("*", mode="after")
    @classmethod
    def check_non_negative(cls, v: float) -> float:
        if v < 0:
            raise ValueError("Optimization objective weights must be non-negative.")
        return v


class AIPlanningRequest(BaseModel):
    planning_date: datetime = Field(..., description="Target planning date")
    horizon: Optional[str] = Field(default="DAILY", description="DAILY, WEEKLY, or MONTHLY")
    corridor_ids: Optional[List[str]] = Field(default=None, description="Target corridor UUIDs (empty for all)")
    departments: Optional[List[str]] = Field(default=None, description="Department filters (ENGINEERING, SIGNAL_TELECOM, TRACTION)")
    max_block_duration_minutes: Optional[int] = Field(default=180, description="Maximum possession duration in minutes")
    min_priority: Optional[float] = Field(default=0.0, description="Minimum asset criticality or priority threshold")
    include_overdue: Optional[bool] = Field(default=True, description="Include overdue maintenance tasks")
    include_critical: Optional[bool] = Field(default=True, description="Include critical priority tasks")
    include_shared_blocks: Optional[bool] = Field(default=True, description="Enable multi-department shared possession bundling")
    optimization_objective: Optional[Dict[str, float]] = Field(
        default={
            "asset_availability": 40.0,
            "maintenance_priority": 25.0,
            "train_impact": 20.0,
            "block_utilization": 15.0
        },
        description="Configurable optimization weights (must sum to 100%)"
    )

    @field_validator("optimization_objective", mode="after")
    @classmethod
    def validate_weights_total(cls, v: Optional[Dict[str, float]]) -> Optional[Dict[str, float]]:
        if v:
            total = sum(v.values())
            if abs(total - 100.0) > 0.5:
                raise ValueError(f"Optimization objective weights must total 100%. Current total: {total:.1f}%")
        return v


class RecommendedBlockItem(BaseModel):
    block_id: str
    corridor_id: str
    corridor_name: str
    date: str
    start_time: str
    end_time: str
    duration_minutes: int
    departments: List[str]
    is_shared_block: bool
    tasks: List[Dict[str, Any]]
    task_count: int
    critical_task_count: int
    affected_trains: List[Dict[str, Any]] = []
    expected_train_delay: float
    maximum_train_delay: float
    asset_availability_gain: float
    block_utilization: float
    optimization_score: float
    confidence: float
    risk_level: str
    reason: str
    alternatives: List[Dict[str, Any]] = []
    constraints_checked: List[str] = []
    approval_status: str


class UnplannedTaskItem(BaseModel):
    task_id: str
    task_code: str
    department: str
    priority: str
    reason: str


class PlanningSummaryData(BaseModel):
    planning_run_id: str
    planning_date: str
    planning_horizon: str
    corridors_analyzed: int
    tasks_analyzed: int
    tasks_selected: int
    tasks_unplanned: int
    critical_tasks_total: int
    critical_tasks_covered: int
    overdue_tasks_covered: int
    blocks_generated: int
    shared_blocks_generated: int
    departments_coordinated: int
    expected_train_delay_minutes: float
    optimization_score: float
    planning_confidence: float
    time_saved_minutes: int
    downtime_reduction_pct: float
    validation_status: str
    solver_duration_seconds: float


class PlanningExplanationData(BaseModel):
    why_selected: List[str] = []
    why_this_time: Optional[str] = ""
    why_not_others: List[str] = []
    overall_narrative: List[str] = []
    validation_checks: List[str] = []


class AIPlanningData(BaseModel):
    planning_run_id: str
    status: str
    planning_date: str
    horizon: str
    corridor_id: str
    corridor_name: str
    summary: PlanningSummaryData
    recommended_blocks: List[RecommendedBlockItem] = []
    unplanned_tasks: List[UnplannedTaskItem] = []
    plan_comparison: Dict[str, Any] = {}
    alternatives: List[Dict[str, Any]] = []
    explanation: PlanningExplanationData


class AIPlanningResponse(BaseModel):
    success: bool = True
    data: AIPlanningData
    message: str = "AI block plan generated successfully"

File: COA software/frontend/src/types/aiPlanner.ts
export interface OptimizationObjectiveWeights {
  asset_availability: number
  maintenance_priority: number
  train_impact: number
  block_utilization: number
}

export interface AIPlanningRequestPayload {
  planning_date: string
  horizon?: 'DAILY' | 'WEEKLY' | 'MONTHLY'
  corridor_ids?: string[]
  departments?: string[]
  max_block_duration_minutes?: number
  min_priority?: number
  include_overdue?: boolean
  include_critical?: boolean
  include_shared_blocks?: boolean
  optimization_objective?: OptimizationObjectiveWeights
}

export interface RecommendedBlock {
  block_id: string
  corridor_id: string
  corridor_name: string
  date: string
  start_time: string
  end_time: string
  duration_minutes: number
  departments: string[]
  is_shared_block: boolean
  tasks: Array<{
    task_id: string
    task_code: string
    department: string
    asset_id: string
    asset_name?: string
    priority: string
    duration_minutes: number
    is_overdue: boolean
    description?: string
  }>
  task_count: number
  critical_task_count: number
  affected_trains: any[]
  expected_train_delay: number
  maximum_train_delay: number
  asset_availability_gain: number
  block_utilization: number
  optimization_score: number
  confidence: number
  risk_level: string
  reason: string
  alternatives: any[]
  constraints_checked: string[]
  approval_status: string
}

export interface UnplannedTask {
  task_id: string
  task_code: string
  department: string
  priority: string
  reason: string
}

export interface PlanningSummary {
  planning_run_id: string
  planning_date: string
  planning_horizon: string
  corridors_analyzed: number
  tasks_analyzed: number
  tasks_selected: number
  tasks_unplanned: number
  critical_tasks_total: number
  critical_tasks_covered: number
  overdue_tasks_covered: number
  blocks_generated: number
  shared_blocks_generated: number
  departments_coordinated: number
  expected_train_delay_minutes: number
  optimization_score: number
  planning_confidence: number
  time_saved_minutes: number
  downtime_reduction_pct: number
  validation_status: string
  solver_duration_seconds: number
}

export interface PlanningExplanation {
  why_selected: string[]
  why_this_time?: string
  why_not_others: string[]
  overall_narrative: string[]
  validation_checks: string[]
}

export interface AIPlanningResultData {
  planning_run_id: string
  status: 'COMPLETED' | 'NO_FEASIBLE_PLAN' | 'FAILED' | string
  planning_date: string
  horizon: string
  corridor_id: string
  corridor_name: string
  summary: PlanningSummary
  recommended_blocks: RecommendedBlock[]
  unplanned_tasks: UnplannedTask[]
  plan_comparison: any
  alternatives: any[]
  explanation: PlanningExplanation
}
